src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,600 @@
+          subscription, topic, PushConfig.getDefaultInstance(), ackDeadline);
+        return;
+
+      } catch (ApiException ae) {
+
+        lastApiException = ae;
+
+        //If the subscription already exists, ignore the error.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.ALREADY_EXISTS)) {
+          return;
+        }
+
+        //This error is thrown is the Topic does not exist.
+        // Call the stop method so the pipeline fails.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+          String message = String.format(MISSING_TOPIC_ERROR_MSG, topic, project);
+          stop(message, ae);
+          return;
+        }
+
+        //Retry if the exception is retryable.
+        if (isApiExceptionRetryable(ae)) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+          continue;
+        }
+
+        //Report that we were not able to create the subscription and stop the receiver.
+        stop(String.format(CREATE_SUBSCRIPTION_ERROR_MSG, subscription), ae);
+        return;
+      } catch (IOException ioe) {
+        //Report that we were not able to create the subscription admin client and stop the receiver.
+        stop(String.format(CREATE_SUBSCRIPTION_ADMIN_CLIENT_ERROR_MSG, subscription), ioe);
+        return;
+      }
+    }
+
+    //If we were not able to create the subscription after 5 attempts, stop the pipeline and report the error.
+    stop(String.format(CREATE_SUBSCRIPTION_RETRY_ERROR_MSG, subscription), lastApiException);
+  }
+
+  /**
+   * Create a new Subscriber Client.
+   *
+   * This method stops the receiver is an exception is thrown when creating the subscriber client.
+   */
+  @Nullable
+  public SubscriberStub createSubscriberClient() {
+    if (isStopped()) {
+      return null;
+    }
+
+    try {
+      return buildSubscriberClient();
+    } catch (IOException ioe) {
+      //This exception is thrown when the subscriber could not be created.
+      //Report the exception and stop the receiver.
+      String message =
+        String.format(SUBSCRIBER_ERROR_MSG, subscription, project);
+      stop(message, ioe);
+    }
+
+    return null;
+  }
+
+  /**
+   * Schedule tasks
+   */
+  public void scheduleTasks() {
+    if (!this.isStopped()) {
+      executor.scheduleAtFixedRate(this::updateMessageRateAndFillBucket, 0, 1, TimeUnit.SECONDS);
+      executor.scheduleWithFixedDelay(this::receiveMessages, 100, 100, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * Fetch new messages for our subscription.
+   * Implements exponential backoff strategy when a retryable exception is received.
+   * This method stops the receiver if a non retryable ApiException is thrown by the Google Cloud subscriber client.
+   */
+  protected void receiveMessages() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+
+    //Try with backoff until stopped or the task succeeds.
+    while (!isStopped()) {
+      try {
+        fetchAndAck();
```

Comment:
if bucket.get() == 0, will we spin in a tight loop for up to a second until the bucket is refilled?

Reply:
This does not really "spin". If `bucket` has a value of 0, the `fetchAndAck` method will return before making any network calls and this method will sleep for 100 ms until the next iteration is invoked.